backends/ubpf/ubpfTable.cpp

```diff
@@ -0,0 +1,408 @@
+*/
+
+#include "ubpfTable.h"
+#include "ubpfType.h"
+#include "ubpfParser.h"
+#include "ir/ir.h"
+#include "frontends/p4/coreLibrary.h"
+#include "frontends/p4/methodInstance.h"
+#include "ubpfControl.h"
+
+namespace UBPF {
+
+    namespace {
+
+        class UbpfActionTranslationVisitor : public EBPF::CodeGenInspector {
+        protected:
+            const UBPFProgram *program;
+            const IR::P4Action *action;
+            cstring valueName;
+
+        public:
+            UbpfActionTranslationVisitor(cstring valueName, const UBPFProgram *program) :
+                    EBPF::CodeGenInspector(program->refMap, program->typeMap), program(program),
+                    action(nullptr), valueName(valueName) {
+                CHECK_NULL(program);
+            }
+
+            bool preorder(const IR::PathExpression *expression) override {
+                auto decl = program->refMap->getDeclaration(expression->path, true);
+                if (decl->is<IR::Parameter>()) {
+                    auto param = decl->to<IR::Parameter>();
+                    bool isParam = action->parameters->getParameter(param->name) == param;
+                    if (isParam) {
+                        builder->append(valueName);
+                        builder->append("->u.");
+                        cstring name = EBPF::EBPFObject::externalName(action);
+                        builder->append(name);
+                        builder->append(".");
+                        builder->append(expression->path->toString());  // original name
+                        return false;
+                    }
+                }
+
+                visit(expression->path);
+                return false;
+            }
+
+            bool preorder(const IR::MethodCallExpression *expression) override {
+                auto mi = P4::MethodInstance::resolve(expression, refMap,
+                                                      typeMap);
+                auto ef = mi->to<P4::ExternFunction>();
+                if (ef != nullptr) {
+                    if (ef->method->name.name == program->model.drop.name) {
+                        builder->append("pass = false");
+                        return false;
+                    } else if (ef->method->name.name == program->model.pass.name) {
+                        builder->append("pass = true");
+                        return false;
+                    } else if (ef->method->name.name ==
+                               program->model.ubpf_time_get_ns.name) {
+                        builder->emitIndent();
+                        builder->append(program->model.ubpf_time_get_ns.name + "()");
+                        return false;
+                    }
+                }
+                program->control->codeGen->preorder(expression);
+            }
+
+            void convertActionBody(const IR::Vector<IR::StatOrDecl> *body) {
+                for (auto s : *body) {
+                    if (!s->is<IR::Statement>()) {
```

Comment:
usually you just need to iterate visit over all statements and write a method for each statement kind.
that's the point of the visitor pattern.
If you will support if statements in actions then this code will be wrong.

Reply:
I refactored a little bit this method, so that it visits all the statements now. We based the implementation of this method based on the `convertActionBody()` from `action.cpp` (simple_switch).